doc/config.md

```diff
@@ -1060,6 +1060,20 @@ they are logged but ignored otherwise.
 Set the pool mode specific to this database. If not set,
 the default `pool_mode` is used.
 
+### load_balance_hosts
+
+When a comma-separated list is specified in `host`, `load_balance_hosts` controls
+which entry is chosen for a new connection.
```

Comment:
It would be great if this same logic would be used when receiving multiple IP adresses from DNS. The easiest way to test this is to create lines in `/etc/host`. Libpq its load_balance_hosts version does this too. So if we don't implement this, we should at least mention in the docs that this setting currently doesn't impact DNS based load balancing.

Reply:
Yeah. I'm wondering if this feature also includes #448 or at least to have it in mind.

Reply:
Interesting, I didn't see that PR before (I guess I didn't go more than 2 years into PR history when joining the project). Right now #448 is completely distinct from this PR:
1. This PR does not implement the `disable` mode for DNS load balancing. It still uses round robin even if `disable` is used.
2. This PR does not implement the `random` mode. PR #448 implements `random` mode for DNS load balancing only, not for host lists (certainly because host lists were not available in PgBouncer in the time that PR was written).

Th design of the config option would allow for implementing both 1 and 2 though. I think 2 isn't required to implement for this PR. A completely missing option isn't very confusing. But I think 1 is quite confusing because the existing option does not apply to all the places that you would expect, so preferably this PR would implement 1 too. If that's hard for some reason, then I think a note in the docs is also sufficient.

Reply:
Hey @JelteF,

I work with the author and he's been a bit too busy to finish responding to your feedback so I'd like to take a stab at it. 

I don't quite understand 1. Maybe I don't understand pgbouncer internals well enough yet but can you give me some pointers on implementing the `disable` mode for DNS load balancing? Sorry for such a vague ask for clarification but I do want to complete this feature as soon as I can.

Reply:
So to clarify: PgBouncer does round-robin load balancing in two ways:
1. If you have multiple hostnames in the `host` key, then it will round robin across these.
2. If a single DNS entry returns multiple IPs, then it load balanaces using round-robin across these IPs: https://www.pgbouncer.org/faq.html#how-to-load-balance-queries-between-several-servers

This same thing is true for `libpq` and `load_balance_hosts` can be used to control the behaviour for both of these there. The implementation of `load_balance_hosts` in this PR only controls the behaviour of 1, not 2. And I think it should control both.

Reply:
Aha thank you for the clarification. I'll look at working up 2.

Reply:
@JelteF I re-based this and took a look but I was wondering: is it possible to merge this with a prominent disclaimer in the docs as you suggested here:

> If that's hard for some reason, then I think a note in the docs is also sufficient.

I would strongly prefer implementing this in a way that's more congruent with libpq and so I would commit to working on a follow-on PR that does that.

The reason I'm so interested in merging this as soon as possible is because:

1. We've been using this and the `target_session_attrs` patch mentioned above in production for some time now. I'm maintaining that and...
2. I detest running an internal fork
3. I'm working through an improvement to our `target_session_attrs` but quite honestly I don't want to change even our internal patch without trying to get it upstreamed
4. Essentially, I want your feedback on my improvements to `target_session_attrs` but to do that we need to get your feedback on our current implementation :)
5. PR-ing that patch doesn't make sense to me without this being incorporated first

Let me know! https://github.com/cosgroveb/pgbouncer/tree/host-strategy-cos-rebased

Reply:
> is it possible to merge this with a prominent disclaimer in the docs as you suggested here

I think it's acceptable, but it would be nice if you could explain why it's not easy to implement it.

Reply:
Hey @JelteF I think that I'm finding this to be a a bit of a yak-shave to start because to test it properly I feel that I need to add a DNS server to the test infrastructure. It's _doable_ don't get me wrong but it does make it difficult to start test-driving the work IMHO.

Reply:
It's actually pretty easy to test this without setting up a full featured DNS server. You can add the same hostname multiple times to the `/etc/hosts` file. i.e. if you add the following there and connect to `myhost` pgbouncer should load balance across 127.0.0.1 and 127.0.0.2
```
127.0.0.1   myhost
127.0.0.2   myhost
```

This trick is actually used by the Postgres tests for `load_balance_hosts` (but to be clear, no automated tests would be needed for PgBouncer for that in this PR).

Reply:
That's not standard behavior for /etc/hosts. Interesting

Reply:
Afaict it's standard behaviour for `getaddrinfo` to return both IP addresess. The DNS level load balancing is then implemented by PgBouncer over this list of IP addresses that it receives (based on my reading of the Pgbouncer docs, i've never tested this feature myself). This is also how `load_balance_hosts` is implemented by libpq.

Reply:
I was mistaken. Looks like `getent hosts XX` returns all IPs listed in /etc/hosts for XX.

Reply:
OK then the only thing that makes this hard @JelteF is my own understanding. Is this where the DNS rr needs to be disabled? https://github.com/cosgroveb/pgbouncer/blob/host-strategy-cos-rebased/src/objects.c#L1647-L1649

Reply:
Okay while `getent hosts XX` returns all the IPs listed in /etc/hosts should I expect to see all the IPs in /etc/hosts for my database host? [I put together this demonstration](https://github.com/pgbouncer/pgbouncer/commit/7b1ff7e69e8906699d94972905dc45b02fd9e42e):

```python
def test_load_balance_hosts_disable_with_dns(bouncer, pg):
    bouncer.default_db = "dns_load_balance_hosts_disable"
    hosts = f"""
127.0.0.53       dnsdbhost
127.0.0.54       dnsdbhost
127.0.0.55       dnsdbhost
    """

    # I have modified utils.py to set listen_addresses='*' for this
    #  demonstrationsbut would use pg.configure/pg.reload for a real test
    with bouncer.run_with_appended_etc_hosts(hosts):
        subprocess_result = capture(
            ["getent", "hosts", "dnsdbhost"],
        )
        getent_result = subprocess_result.split("\n")
        assert "127.0.0.53      dnsdbhost" == getent_result[0]
        assert "127.0.0.54      dnsdbhost" == getent_result[1]
        assert "127.0.0.55      dnsdbhost" == getent_result[2]

        bouncer.sql(query="SELECT 1", user="bouncer", password="zzzz", dbname="dns_load_balance_hosts_disable")
        with bouncer.admin_runner.cur() as cur:
            results = cur.execute("SHOW DNS_HOSTS").fetchall()
            result = [r for r in results if r[0] == 'dnsdbhost'][0]

            assert result[2] != "127.0.0.53:0"
```

```====================================================== FAILURES =======================================================
______________________________________ test_load_balance_hosts_disable_with_dns _______________________________________
/home/admin/code/pgbouncer/test/test_load_balance_hosts.py:70: in test_load_balance_hosts_disable_with_dns
    assert result[2] != "127.0.0.53:0"
E   AssertionError: assert '127.0.0.53:0' != '127.0.0.53:0'
        bouncer    = <test.utils.Bouncer object at 0x7f0499fbb2e0>
        cur        = <psycopg.Cursor [closed] [BAD] at 0x7f049b445460>
        getent_result = ['127.0.0.53      dnsdbhost', '127.0.0.54      dnsdbhost', '127.0.0.55      dnsdbhost', '127.0.0.53      dnsdbhost', '127.0.0.54      dnsdbhost', '127.0.0.55      dnsdbhost', ...]
        hosts      = '\n127.0.0.53       dnsdbhost\n127.0.0.54       dnsdbhost\n127.0.0.55       dnsdbhost\n    '
        pg         = <test.utils.Postgres object at 0x7f0499fbbac0>
        result     = ('dnsdbhost', 14, '127.0.0.53:0')
        results    = [('dnsdbhost', 14, '127.0.0.53:0')]
        subprocess_result = '127.0.0.53      dnsdbhost\n127.0.0.54      dnsdbhost\n127.0.0.55      dnsdbhost\n127.0.0.53      dnsdbhost\n127.0.0.54      dnsdbhost\n127.0.0.55      dnsdbhost\n'
```

Because the work on the related feature, `target_session_attrs` is substantially complete and in need of review I do favor merging this branch as soon as possible. As you can see, the testing situation is a _bit icky_ at the moment for the DNS sub-feature of load_balance_hosts but with some time I can complete it. It would be super beneficial to me if we can start working on getting `target_session_attrs` and I think it would be to other users as well.

Reply:
Sounds reasonable to me.

Reply:
I'll get push access to @dpirotte's fork, push up a more recent rebase, and address your docs feedback below `it is recommended to set server_login_retry lower than the default`